index.js

```diff
@@ -370,34 +370,33 @@ async function fastifyStatic (fastify, opts) {
       for (const [dirname, rootPath] of indexDirs.entries()) {
         const pathname = dirname + (dirname.endsWith('/') ? '' : '/')
         const file = '/' + pathname.replace(prefix, '')
-        setUpHeadAndGet(fastify, routeOpts, pathname, file, rootPath)
+        setUpHeadAndGet(routeOpts, pathname, file, rootPath)
 
         if (opts.redirect === true) {
-          setUpHeadAndGet(fastify, routeOpts, pathname.replace(/\/$/, ''), file.replace(/\/$/, ''), rootPath)
+          setUpHeadAndGet(routeOpts, pathname.replace(/\/$/, ''), file.replace(/\/$/, ''), rootPath)
         }
       }
     }
   }
 
-  function setUpHeadAndGet (fastify, routeOpts, route, file, rootPath) {
+  function setUpHeadAndGet (routeOpts, route, file, rootPath) {
     const toSetUp = {
       ...routeOpts,
       method: ['HEAD', 'GET'],
       url: route,
-      handler: serveFileHandler
+      handler: function serveFileHandler (req, reply) {
+        const file = req.routeConfig.file
+        const rootPath = req.routeConfig.rootPath
+        pumpSendToReply(req, reply, file, rootPath)
+      }
     }
     toSetUp.config = toSetUp.config || {}
     toSetUp.config.file = file
     toSetUp.config.rootPath = rootPath
     fastify.route(toSetUp)
   }
-
```

Comment:
imho by integratin the serveFileHandler into setUpHeadAndGet you increased the memory consumption of this plugin.

Reply:
Thanks for the comment, this is a recurring question I keep asking myself

Normally wrapped functions are a bad idea but here, since the plugin function runs once and sets everything up, I really can't tell if it's better to extract most functions out, or just move things inside the plugin function to remove the need to pass things like the `fastify` instance limit the scope

What do you think?

Reply:
I quiet dont get your question. I agree with you, that we dont need to pass the fastify instance to a function, if we dont call it. 

Regarding the serveFileHandler, you can see that it will be called by potentially alot of routes. Also serveFileHandler does not modify anything regarding the plugin or the fastify instance. 

Its worth to extract functions, when you know you assign them multiple times and also I would say where you know that the input or the output wont be polymorphic.

Reply:
Yeah yeah, you're right in this case. Let me give an example:

```js
function checkPath (fastify, rootPath) {
  if (typeof rootPath !== 'string') {
    throw new Error('"root" option must be a string')
  }
  if (path.isAbsolute(rootPath) === false) {
    throw new Error('"root" option must be an absolute path')
  }

  let pathStat

  try {
    pathStat = statSync(rootPath)
  } catch (e) {
    if (e.code === 'ENOENT') {
      fastify.log.warn(`"root" path "${rootPath}" must exist`)
      return
    }

    throw e
  }

  if (pathStat.isDirectory() === false) {
    throw new Error('"root" option must point to a directory')
  }
}
```
Even though it doesn't matter *that* much, functions like this can be moved inside `fastifyStatic` so that there is no need to pass `fastify` around

Reply:
Maybe it is easier to test them, if they are not totally encapsulated.